include/GafferScene/EditScopeAlgo.h

```diff
@@ -51,12 +51,22 @@ namespace GafferScene
 namespace EditScopeAlgo
 {
 
+// 'readOnlyReason' methods
+// ========================
+// If is often necessary to determine the cause of the read-only state of an
+// edit, or whether an edit can be added to any given scope. These methods
+// return the outward-most GraphComponent that is causing any given edit (or
+// potential edit creation) to be read-only. Tools that create edits within a
+// scope should first check this returns null before calling any 'acquire'
+// method to avoid incorrectly modifying locked nodes/plugs.
+
 // Pruning
 // =======
 
 GAFFERSCENE_API void setPruned( Gaffer::EditScope *scope, const ScenePlug::ScenePath &path, bool pruned );
 GAFFERSCENE_API void setPruned( Gaffer::EditScope *scope, const IECore::PathMatcher &paths, bool pruned );
 GAFFERSCENE_API bool getPruned( Gaffer::EditScope *scope, const ScenePlug::ScenePath &path );
+GAFFERSCENE_API const Gaffer::GraphComponent *pruneReadOnlyReason( const Gaffer::EditScope *scope, const ScenePlug::ScenePath &path );
```

Comment:
I don't think `path` has any relevance here does it? I also wonder if `pruneReadOnlyReason()` sounds as if it might prune something? Is there a better name? `pruningReadOnlyReason()` perhaps?

I also went looking to see where we might be able to use this, and found the following code and todos in EditScopeUI.py :

```
	## \todo Accessing the processor directly like this rather defeats the object of
	# the EditScopeAlgo API. Consider additional API to say whether or not a user
	# should be able to edit something, based on all the `readOnly` and `enabled` checks
	# we're performing manually here.
	pruningProcessor = editScope.acquireProcessor( "PruningEdits", createIfNecessary = False )
	if pruningProcessor is not None and Gaffer.MetadataAlgo.readOnly( pruningProcessor["paths"] ) :
		return True

	with viewer.getContext() :
		if not editScope["enabled"].getValue() :
			# Spare folks from deleting something when it won't be
			# apparent what they've done until they reenable the
			# EditScope.
			return True
                if pruningProcessor is not None and not pruningProcessor["enabled"].getValue() :
			return True
```

So I think we are right to be introducing some sort of helper API, but I'm now worried that it should perhaps be taking into account the enabled status of processors as well. Any thoughts?

Reply:
- Agreed off line to rename to `prunedReadOnlyReason`, keeping `path` for consistency/not leaking implementation (see [here](https://github.com/GafferHQ/gaffer/pull/3840/commits/28f7be251a066a9485e11e567dd566be35ff8e55)).
 - Will update `EditScopeUI` in a separate PR to use new methods.

Reply:
Subsequent to further thought [this](https://github.com/GafferHQ/gaffer/pull/3840/commits/675c04354d917c211eb4a9c0c401590615d2ca0f) removes `paths` (the polymorphic nature of `setPruned` makes it less helpful, and pruned is an outlier in terms of API anyway), and [this](https://github.com/GafferHQ/gaffer/pull/3840/commits/0932124264d16abee5fd4882bfb7eddf0648350c) updates `EditScopeUI`. We still need to fix the transform tools, et al.